okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt

```diff
@@ -0,0 +1,158 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.ws
+
+import java.io.IOException
+import java.net.ProtocolException
+import okhttp3.Response
+
+data class WebSocketOptions(
+  @JvmField val compressionEnabled: Boolean,
+  @JvmField val contextTakeover: Boolean
+) {
+  companion object {
+    private const val HEADER_WEB_SOCKET_EXTENSION = "Sec-WebSocket-Extensions"
+    private const val EXTENSION_PERMESSAGE_DEFLATE = "permessage-deflate"
+
+    private val NO_COMPRESSION = WebSocketOptions(
+      compressionEnabled = false,
+      contextTakeover = false
+    )
+
+    private val COMPRESSION_NO_TAKEOVER = WebSocketOptions(
+      compressionEnabled = true,
+      contextTakeover = false
+    )
+
+    private val COMPRESSION_WITH_TAKEOVER = WebSocketOptions(
+      compressionEnabled = true,
+      contextTakeover = true
+    )
+
+    @Throws(IOException::class)
+    @JvmStatic
+    fun parseServerResponse(response: Response): WebSocketOptions {
+      // No extension header - server does not support permessage-deflate.
+      val header = response.header(HEADER_WEB_SOCKET_EXTENSION) ?: return NO_COMPRESSION
+
+      // Server is free to return empty header to indicate no compression
+      // See end of https://tools.ietf.org/html/rfc7692#section-5 chapter
+      if (header.isEmpty()) {
+        return NO_COMPRESSION
+      }
+
+      val extension = header
+        .split(", ")
+        // Additional or fallback extensions are not supported currently.
+        // See https://tools.ietf.org/html/rfc7692#section-5.2 for details.
+        .getOrNull(0)
+
+      if (extension.isNullOrBlank()) {
+        throw ProtocolException("$HEADER_WEB_SOCKET_EXTENSION malformed: '$header'")
+      }
+
+      val tokens = extension.split("; ").map { it.replace(";", "") }
+      if (tokens.isEmpty()) {
+        throw ProtocolException("Extension could not be parsed: '$extension'")
+      }
+
+      val extensionName = tokens[0]
+      if (extensionName != EXTENSION_PERMESSAGE_DEFLATE) {
+        // Client MUST fail for extension it did not ask for
+        throw ProtocolException("Extension not supported: '$extensionName'")
+      }
+
+      val options = tokens.drop(1).map(Option.Companion::parse)
+      if (options.isEmpty()) {
+        // Server did not force client_no_context_takeover,
+        // so client can use compression with context takeover.
+        return COMPRESSION_WITH_TAKEOVER
+      }
+
+      if (options.toSet().size != options.size) {
+        // Client MUST fail for duplicate options.
+        throw ProtocolException("Duplicate options found in '$extension'")
+      }
+
+      val clientNoContextTakeover = Option.CLIENT_NO_CONTEXT_TAKEOVER in options
```

Comment:
There’s also an independent field, server_no_context_takeover, that we need to handle when inflating